### ISSUE TITLE
[XLA:TPU] Identity HLO transform is rejected by HloVerifier for valid multi-slice sharding

### ISSUE THREAD

**Issue by hhhhsdxxxx** [openxla/xla#47778]
## Summary

On a two-slice TPU compile topology, registering a pre-scheduler HLO transform
that returns the input `HloModuleProto` bytes unchanged causes XLA's
post-transform `HloVerifier` to reject a valid global sharding.

The same program compiles successfully when the hook is not registered. No HLO
instruction, sharding, or module field needs to be changed to reproduce the
failure; returning the original bytes (and therefore reporting `changed=true`)
is sufficient.

## Minimal reproduction

```python
import jax
import jax.numpy as jnp
import numpy as np
from jax.experimental.topologies import get_topology_desc
from jax.extend import xla as jax_xla
from jax.sharding import Mesh, NamedSharding, PartitionSpec as P


def identity_changed(proto_bytes: bytes) -> bytes:
  # Deliberately return exactly the bytes supplied by XLA.
  return proto_bytes


jax_xla.register_hlo_module_transformation(
    identity_changed,
    name="identity_changed_multislice_repro",
    stage=jax_xla.PipelineStage.PRE_SCHEDULER,
    platforms=("tpu",),
)

topology = get_topology_desc(
    platform="tpu",
    topology_name="tpu7x:4x8x8",
    chip_config_name="default",
    chips_per_host_bounds=(2, 2, 1),
    num_slices=2,
    wrap=(True, True, True),
)
devices = list(topology.devices)
mesh = Mesh(
    np.asarray(devices, dtype=object).reshape(2, 512),
    ("dcn", "fsdp"),
)
sharding = NamedSharding(mesh, P("fsdp", None))
argument = jax.ShapeDtypeStruct((512, 8), jnp.bfloat16, sharding=sharding)


def step(value):
  return value + jnp.asarray(1, dtype=value.dtype)


jax.jit(
    step,
    in_shardings=sharding,
    out_shardings=sharding,
).lower(argument).compile()
```

Removing the `register_hlo_module_transformation` call makes this compile
successfully.

The HLO proto supplied to the callback has a per-slice device assignment:

```text
device_assignment.replica_count      = 1
device_assignment.computation_count  = 512
```

while the parameter has a global two-slice replicated sharding:

```text
tile_assignment_dimensions = (512, 1, 2)
tile assignment device count = 1024
last_tile_dim_replicate = true
```

## Actual behavior

```text
INVALID_ARGUMENT: during context [Unknown]: Invalid sharding for instruction:
%param = bf16[1,8]{1,0:T(2,128)(2,1)} parameter(0),
sharding={devices=[512,1,2]<=[2,512]T(1,0) last_tile_dim_replicate},
metadata={op_name="value"}: tile_assignment should have 512 devices but has 1024

Note: While validating sharding
{devices=[512,1,2]<=[2,512]T(1,0) last_tile_dim_replicate}
against shape bf16[1,8]
```

## Expected behavior

An identity transform should preserve the successful behavior of the unhooked
compile. In particular, the verifier used after the transform should accept the
TPU compiler's pre-scheduler representation containing a per-slice device
assignment and a global multi-slice replicated sharding.

## Suspected root cause

`ApplyXlaTransforms::RunImpl` runs a default `HloVerifier` whenever a transform
returns `changed=true`:

https://github.com/openxla/xla/blob/06fceb446601f318a01259662e15e13df6256c17/xla/service/xla_transform.cc#L143-L150

The generic `InstructionVerifier` derives its sharding device count solely from
`module->config().num_partitions()`:

https://github.com/openxla/xla/blob/06fceb446601f318a01259662e15e13df6256c17/xla/service/hlo_verifier.h#L393-L402

At the TPU pre-scheduler hook stage, this is 512 after the compiler has localized
the device assignment to one slice, while some valid global shardings still
describe all 1024 devices. The normal TPU compilation pipeline accepts this
intermediate representation, but the additional generic verifier does not.

A possible localized workaround would be to retain the structural HLO
verification after user transforms while disabling only sharding device-number
verification at this TPU pipeline stage. A more complete fix may need to retain
the global device-count context when validating these multi-slice shardings.

## System information

```text
platform: TPU
compile topology: tpu7x:4x8x8, 512 logical devices per slice, 2 slices
logical mesh: dcn=2, fsdp=512
JAX: 0.11.0
jaxlib: 0.11.0
Python: 3.12.12 (GCC 12.2.0)
OS: Linux 6.12.85+, x86_64, glibc 2.36
```

The same failure was reproduced with all of these `libtpu` releases:

```text
0.0.44
0.0.44.1
0.0.46 (latest published release at the time of filing)
```

The relevant verifier logic is also still present at public OpenXLA `main`
commit `06fceb446601f318a01259662e15e13df6256c17` (2026-08-24).

**Comment by akuegel** [comment]
It seems we would need some mechanism for allowing to configure the ApplyXlaTransforms step with verifier metadata, or otherwise allowing to configure the verifier to be used. @andydavis1 I assigned this issue to you as it seems you worked on this feature to allow registering HLO transformation passes.
